inference/server/main.py

```diff
@@ -166,63 +142,62 @@ async def event_generator():
 
                 yield {
                     "retry": settings.sse_retry_timeout,
-                    "data": TokenResponseEvent(token=response_packet.token).json(),
+                    "data": interface.TokenResponseEvent(token=response_packet.token).json(),
                 }
-            logger.info(f"Finished streaming {chat.id} {len(result_data)=}")
+            logger.info(f"Finished streaming {id} {len(result_data)=}")
         except Exception:
-            logger.exception(f"Error streaming {chat.id}")
+            logger.exception(f"Error streaming {id}")
+            raise
 
-        chat.conversation.messages.append(
-            protocol.ConversationMessage(
-                text=response_packet.generated_text.text,
-                is_assistant=True,
-            )
-        )
-        chat.pending_message_request = None
+        try:
+            with contextlib.contextmanager(create_session)() as session:
+                chat_repository = create_chat_repository(session)
+                chat_repository.add_assistant_message(id=id, text=response_packet.generated_text.text)
+        except Exception:
+            logger.exception("Error adding assistant message")
 
-    return EventSourceResponse(event_generator())
+    return EventSourceResponse(event_generator(id))
 
 
 @app.websocket("/work")
-async def work(websocket: fastapi.WebSocket):
+async def work(websocket: fastapi.WebSocket, chat_repository: ChatRepository = Depends(create_chat_repository)):
     await websocket.accept()
     worker_config = inference.WorkerConfig.parse_raw(await websocket.receive_text())
     try:
         while True:
-            print(websocket.client_state)
             if websocket.client_state == fastapi.websockets.WebSocketState.DISCONNECTED:
                 logger.warning("Worker disconnected")
                 break
             # find a pending task that matches the worker's config
             # could also be implemented using task queues
             # but general compatibility matching is tricky
-            for chat in CHATS.values():
-                if (request := chat.pending_message_request) is not None:
-                    if chat.message_request_state == MessageRequestState.pending:
-                        if request.compatible_with(worker_config):
-                            break
+            for chat in chat_repository.get_pending_chats():
+                request = chat.pending_message_request
+                if request.compatible_with(worker_config):
+                    break
             else:
-                logger.debug("No pending tasks")
                 await asyncio.sleep(1)
                 continue
 
-            chat.message_request_state = MessageRequestState.in_progress
+            chat_repository.set_chat_state(chat.id, interface.MessageRequestState.in_progress)
```

Comment:
The "dequeue" operation needs to set the chat immediately into a state that prevents others clients from also dequeuing it. Imagine we have 5k connected inference clients and all want to handle chats... this dequeue op will be a massive congestion point. A possible strategy could be to find and update in one sql-statement sent to the db .. or to use a dedicated queue service.

Reply:
yes absolutely, this part will have to be re-worked anyway, probably best to do it with redis, too.
I'll merge this one in and improve later